pandas/core/arrays/arrow/array.py

```diff
@@ -1999,27 +2044,110 @@ def _groupby_op(
-            values = self._to_timedeltaarray()
+        pa_supported_groupby_aggregations = [
+            "any",
+            "all",
+            "sum",
+            "prod",
+            "min",
+            "max",
+            "mean",
+            "sem",
+            "std",
+            "var",
+        ]
+        # aggregations not yet supported natively in pyarrow:
+        #  - first/last: https://github.com/apache/arrow/issues/36709
+        #  - nth
+        #  - median
+
+        if how not in pa_supported_groupby_aggregations:
+            # maybe convert to a compatible dtype for groupby
+            values: ExtensionArray
+            pa_type = self._pa_array.type
+            if pa.types.is_timestamp(pa_type):
+                values = self._to_datetimearray()
+            elif pa.types.is_duration(pa_type):
+                values = self._to_timedeltaarray()
+            else:
+                values = self._to_masked()
+
+            result = values._groupby_op(
+                how=how,
+                has_dropped_na=has_dropped_na,
+                min_count=min_count,
+                ngroups=ngroups,
+                ids=ids,
+                **kwargs,
+            )
+            if isinstance(result, np.ndarray):
+                return result
+            return type(self)._from_sequence(result, copy=False)
+
+        arr = self._values_for_reduction(how)
+
+        pa_name = {
+            "prod": "product",
+            "std": "stddev",
+            "var": "variance",
+        }.get(how, how)
+
+        # pyarrow does not accept a min_count of -1
+        if min_count == -1:
+            if how in ["any", "all"]:
+                min_count = 0
+            else:
+                min_count = 1
+
+        if how in ["std", "var"]:
+            aggregations = [
+                (
+                    "val",
+                    pa_name,
+                    pc.VarianceOptions(ddof=kwargs["ddof"], min_count=min_count),
+                )
+            ]
+        elif how == "sem":
+            aggregations = [
+                (
+                    "val",
+                    "stddev",
+                    pc.VarianceOptions(ddof=kwargs["ddof"], min_count=min_count),
+                ),
+                (
+                    "val",
+                    "count",
+                    pc.CountOptions(mode="only_valid"),
+                ),
+            ]
         else:
-            values = self._to_masked()
-
-        result = values._groupby_op(
-            how=how,
-            has_dropped_na=has_dropped_na,
-            min_count=min_count,
-            ngroups=ngroups,
-            ids=ids,
-            **kwargs,
+            aggregations = [
+                (
+                    "val",
+                    pa_name,
+                    pc.ScalarAggregateOptions(min_count=min_count),
+                )
+            ]
+
+        res = (
+            pa.Table.from_pydict({"id": ids, "val": arr})
+            .group_by("id")
```

Comment:
IIUC this is re-calculating the codes we already have in our GroupBy object.  I see the appeal, but am wary of this pattern.  I think the long-term solution is to get pyarrow to expose something to which we can pass the already-computed codes?

Reply:
Yep, that's fair. Also, it would be nice to pass all aggregations at once to pyarrow - see the last comment in the OP.

Reply:
> Longer term, there are potentially big performance gains if all arrow-backed aggregations can be passed to pyarrow at once. At the moment, we're iterating column by column. Here is a small example showing the impact:

Any idea what is driving the perf impact?  Does the pyarrow version do something in parallel?

Reply:
I can't tell from reading the docs whether anything is done in parallel. It could be simply (as you've pointed out) that pyarrow is only computing the group codes once when done in a single batch and recomputing them when done iteratively.